src/rt/region/region_api.h

```diff
@@ -0,0 +1,332 @@
+// Copyright Microsoft and Project Verona Contributors.
+// SPDX-License-Identifier: MIT
+#pragma once
+
+#include "freeze.h"
+#include "region.h"
+
+
+namespace verona::rt::api
+{
+  namespace internal
+  {
+    class RegionContext
+    {
+      struct RegionFrame
+      {
+        Object* entry_point;
+        RegionBase* region;
+        RegionFrame* prev;
+      };
+
+      RegionFrame* top;
+
+    public:
+      static RegionContext& get_region_context()
+      {
+        static thread_local RegionContext context;
+        return context;
+      }
+
+      static void push(Object* entry_point, RegionBase* region)
+      {
+        RegionFrame* frame = new RegionFrame();
+        frame->entry_point = entry_point;
+        frame->region = region;
+        frame->prev = get_region_context().top;
+        get_region_context().top = frame;
+      }
+
+      static void pop()
+      {
+        auto& t = get_region_context();
+        RegionFrame* frame = t.top;
+        t.top = frame->prev;
+        delete frame;
+      }
+
+      static Object*& get_entry_point()
+      {
+        return get_region_context().top->entry_point;
+      }
+
+      static RegionBase* get_region()
+      {
+        return get_region_context().top->region;
+      }
+    };
+  }
+
+  using namespace internal;
+
+  /**
+   * Check if pointer points to a new region.
+   */
+  inline bool is_region_ref(Object* o)
+  {
+    // Check if iso
+    if (!o->debug_is_iso())
+      return false;
+    // Check for entry point
+    return RegionContext::get_entry_point() != o;
+  }
+
+  /**
+   * Open supplied region, and return entry point.
+   */
+  inline void open_region(Object* r)
+  {
+    auto md = r->get_region();
+    RegionContext::push(r, md);
+    switch (Region::get_type(md))
+    {
+      case RegionType::Trace:
+      case RegionType::Arena:
+        break;
+      case RegionType::Rc:
+        ((RegionRc*)md)->open(r);
+        break;
+      default:
+        abort();
+    }
+  }
+
+  /**
+   * Close current region
+   */
+  inline void close_region()

```

Comment:
This will crash if called without a current region in thread local storage right? Shall we add an assert on entry that the top of the stack is not a nullptr?